### DISCUSSION TITLE
UEFI implementation

### DISCUSSION THREAD

**Discussion by ADev531** [CosmosOS/Cosmos#2649] [Ideas]
I think if use uefi, no graphic limit, no disk limit. Uefi can skip some part ‘Starting ACPI’, and can support 32, 64!

**Comment by selkij** [comment]
This would take too much time to make, IL2CPU adaptation, cosmos adaptation and so much other stuff, its more like a dream to have it

**Comment by ascpixi** [comment]
I don't think that UEFI support in Cosmos is a "dream" away from reality; the Limine bootloader already abstracts the majority of the communication you do with firmware interfaces - it's just that Cosmos still uses the Multiboot2 protocol. Cosmos isn't "bound" to BIOS, since we don't depend on BIOS once we exit 16-bit real mode.

The only issue I can see is porting Cosmos to use bootloader-provided services rather than firmware interface-provided ones. This would mean that the VBE class would be thrown out, and instead the framebuffer provided by the bootloader would be used. Multiboot2 actually provides such a tag - it's the `framebuffer` tag, which Limine also implements. This tag contains the information required for video display, and is firmware-agnostic. It's implementation-defined, but with Limine, it uses VBE on BIOS and GOP on UEFI.

UEFI will not provide 64-bit support; BIOS is not the component that is holding Cosmos back from transitioning into 64-bit long mode. Structures like the GDT would have to be updated (simple), and all pointers would have to be converted to quad-words (not so simple, since Cosmos uses types such as `uint` to represent memory addresses).

However, for now, focus is more on cleaning and refactoring old code. There are major bugs in the current master branch that still need to be resolved before we start using bootloader-provided abstraction layers.

**Reply by ADev531** [reply]
I know.
I made x64 os before using uefi
and now im want cosmos

**Comment by AnErrupTion** [comment]
Wait, Cosmos uses VBE manually and not Multiboot2's framebuffer tag????!!!!

**Reply by ascpixi** [reply]
It does certainly [look like it](https://github.com/CosmosOS/Cosmos/blob/master/source/Cosmos.Core/Multiboot.cs#L154). Tag type 7 corresponds to [VBE information](https://www.gnu.org/software/grub/manual/multiboot2/multiboot.html#VBE-info) - not framebuffer information.

**Reply by ADev531** [reply]
Yay, multiboot has updated to 2.
Now not using vbe information!
but what is efi64?

**Comment by zarlo** [comment]
one day cosmos will move to UEFI but we have a lack of man power

**Reply by tacitmarmot271** [reply]
well i can be a part of the dev team

**Reply by R0fael** [reply]
me too

**Reply by zarlo** [reply]
well PR's are open

**Comment by MishaProductions** [comment]
Solved by https://github.com/CosmosOS/Cosmos/pull/2828

**Reply by ADev531** [reply]
thanks!
